Rewrite ViewModelCache

davecoates opened the pull request:
This implementation is greatly simplified (and about half as long) but the ViewModelCache file has been essentially rewritten - view the whole file rather than trying to read the diff. This resolves some long standing bugs (see test cases that were previously skipped) and makes it easier to work with for any extensions. 

ViewModelFactory has also had large parts rewritten. Notably there is now a base class that is always extended rather than being entirely dynamic. This file might be easier to be viewed in its entirety rather than a diff but a lot of it is old code just moved around.

Other notable changes
- ViewModelFactory no longer supports implicit primary keys. You always must define this yourself.
- You can now specify the cache to use on a relation field (defaults to the ViewModel.cache). This is mainly useful for testing.
- Types now verify field names across relations.
    - These _don’t_ work across circular relations but we can live with that I think
- Types are generally much more comprehensive for caching and factory
- Add [tsd](https://github.com/SamVerschueren/tsd) for typescript testing

davecoates committed:
test: setup tsd

davecoates committed:
fix: viewmodelfactory type improvements

- Improve general field handling
- Improve pk types
- Improve augment types
- Add type tests

davecoates committed:
fix: rewrite viewmodel types & implementation

- Better typing for fields
- Has BaseViewModel class now that everything extends from

davecoates committed:
sys: upgrade babel

primarily to support latest typescript

davecoates committed:
fix: update testbed model def

davecoates committed:
feature: fieldUtils

Contains utility methods and classes for representing field 'paths'
(where a path is either a single field, eg 'name' or array of strings
used to index across a relation, eg. ['group', 'owner', 'name'])

davecoates committed:
feature: RelatedViewModelField 'cache' option

Add ability to specify a different cache to use (defaults to viewModel.cache)

Type improvements

davecoates committed:
feature: rewrite ViewModelCache

This vastly improves the type information, simplifies the implementation and resolves
some long standing bugs (see previously skipped tests)

davecoates committed:
fix: specify pkFieldName

davecoates committed:
sys: upgrade typescript

davecoates committed:
fix: add FieldPaths export

davecoates committed:
fix: doc fixes & minor type fix

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:
```diff
@@ -40,643 +45,195 @@ function withEnableListeners<T>(run: () => T): T {
-     *
-     * Must be kept in sync with `getCacheKey`/`getFieldNameCacheKey`
+     * When a record has relation fields it needs to listen for changes on those records so it can
+     * update itself and notify it's direct listeners. This Map keeps track of those listeners.
```
possible to rewrite this comment? had to read it quite a few times to get what its saying (with that tingling uncertainty)

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:
```diff
@@ -40,643 +45,195 @@ function withEnableListeners<T>(run: () => T): T {
+     * This is used to choose the most recent record when a new record with a subset
+     * of the fields needs to be created. For example if there are 2 existing records
+     * created in this order:
+     * { id: 1, firstName: 'Jon', lastName: 'Doe', active: true }
```
would suggest change the `active` key in this example to sth else as active could have other meanings (ie. its easy for this to be seen as not part of record itself)

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:
```diff
@@ -40,643 +45,195 @@ function withEnableListeners<T>(run: () => T): T {
     }
 
     /**
      * Set a value for the specified key notifying any listeners
```
* Set a value for the specified key_, then_ notify any listeners
?

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:
```diff
@@ -40,643 +45,195 @@ function withEnableListeners<T>(run: () => T): T {
-                    )
-                );
-            }
+        let before = this.cache.get(key) || null;
```
so even if the value is the same, the lastUpdatedKeys order would still change - this could get a mention in doc if its intended

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:
```diff
@@ -40,643 +45,195 @@ function withEnableListeners<T>(run: () => T): T {
-                record = record.record;
-            }
-            return new RecordPointer(this.cache.get(key), record);
+        if (before && value && isShallowEqual(before, value)) {
```
uh, would isShallowEqual return true if one of them is null but the other is not? 
as both null cases are covered above already

davecoates replied:
ah yeah dunno why did this - should just be able to do the isShallowEqual check

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:
```diff
@@ -40,643 +45,195 @@ function withEnableListeners<T>(run: () => T): T {
+        // record doesn't already exist for key - see if it's a subkey of existing entries
+        for (const cacheKey of this.lastUpdatedKeys) {
+            const record = this.cache.get(cacheKey);
+            if (record && cacheKey.isSubset(key, true)) {
```
performance: is the .get heavier or isSubset heavier? asking coz order of execution here's a bit unintuitive

davecoates replied:
`isSubset` i imagine - `get` is just a `Map.get` call

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:
```diff
@@ -40,643 +45,195 @@ function withEnableListeners<T>(run: () => T): T {
+                )) {
+                    const relationField = this.viewModel.getField(
+                        relationFieldName
+                    ) as BaseRelatedViewModelField<any, any, any>;
```
(ugh, whats the line width setting we're using in presto..)

davecoates replied:
printWidth of 100 - same as all our projects :P

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:
```diff
@@ -40,643 +45,195 @@ function withEnableListeners<T>(run: () => T): T {
+                    if (idOrIds == null || (Array.isArray(idOrIds) && idOrIds.length === 0)) {
+                        data[relationFieldName] = relationField.many ? [] : null;
+                    } else {
+                        if (Array.isArray(idOrIds)) {
```
it feels like the better check to use is `relationField.many`. are there cases that they'll  otherwise still be an array?

davecoates replied:
nup - will change

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:
```diff
@@ -686,94 +243,153 @@ class RecordCache<ViewModelClassType extends ViewModelConstructor<any, any, any>
+                relationFieldName
+            ) as BaseRelatedViewModelField<any, any, any>;
+            const idOrIds = baseRecord[relationField.sourceFieldName];
+            if (idOrIds != null) {
```
minor: this's similar to one above so it might make sense to line up the logic? (ie. check idOrIds being empty array at this level & set newData[] as so to avoid hitting getList down below)

davecoates replied:
I've reversed the condition:

```
if (idOrIds == null || (Array.isArray(idOrIds) && idOrIds.length === 1)) {
    newData[relationFieldName] = relationField.many ? [] : null;
} else {
    ...
}
````

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:
```diff
@@ -686,94 +243,153 @@ class RecordCache<ViewModelClassType extends ViewModelConstructor<any, any, any>
-     * @param listener Function to call with any changes
+     * Given a `baseRecord` with no related fields resolved (ie. we just have the id's, not the nested records) construct
+     * a record with all the nested records attached if those records exist in the cache.
+     *
```
can use a brief mentioning here like '...if all of those records exist in the cache, otherwise return null'

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:
```diff
@@ -686,94 +243,153 @@ class RecordCache<ViewModelClassType extends ViewModelConstructor<any, any, any>
+            return;
+        }
+        registeredPaths.push(key);
+        this.relationListenerUnsubscribe.get(key)?.forEach(unsub => unsub());
```
why would there? think i might be missing sth - can i get an example?

davecoates replied:
I'll add this comment

```
        // This may already be set for a previous call to this function for a different record (ie.
        // a previous version of the record with different related record id(s)). In that case remove
        // the old listeners and add new ones to ensure they match the latest related record id(s).
```

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:
```diff
@@ -1007,25 +668,59 @@ const defaultListenerBatcher = {
+     */
+    private getPkCacheKey(pk: ExtractPkFieldParseableValueType<ViewModelClassType>): string {
+        if (typeof pk === 'object') {
+            return this.viewModel.pkFieldNames.map(fieldName => pk[fieldName]).join('⁞');
```
😆 👍

fanglong-alliance replied:
looks like ⁞ was defined as a const in fieldUtils - should use here

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelCache.ts:
```diff
@@ -1527,6 +991,78 @@ export default class ViewModelCache<
+        removeNulls?: RemoveNullsT
+    ): RemoveNullsT extends true
+        ? PartialViewModel<ViewModelClassType, T>[]
+        : (PartialViewModel<ViewModelClassType, T> | null)[];
```
my eyes! 🤣
 
perhaps rearrange these 3 to make it a bit clearer? (mainly the 2nd overload)

davecoates replied:
hmm don't see how i can re-arrange it to make any meaningful difference

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts:
```diff
@@ -1,82 +1,306 @@
+    : {
+          [K in keyof FieldMappingType as K extends PkFieldType[number]
+              ? K
+              : never]: FieldMappingType[K];
```
(this's really not pretty🤣 : wth's `:never]:`? )

davecoates replied:
`never` will mean only fields that extend `PkFieldType[number]` will be included

fanglong-alliance replied:
yeah what i mean's prettier did a &^@~&!@ job there (ie a side comment hence its entirely inside a () pair) that it caused the line to begin with `:never]:` which gave me a pause 😆

davecoates replied:
ahhh my bad i thought you meant what does it mean

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts:
```diff
@@ -1,82 +1,306 @@
+                ([relationName, relationFieldNames]) =>
+                    ` The relation '${relationName}' for fields '${relationFieldNames.join(', ')}'`
+            );
+        }
```
would the msg look funny if the only missing parts are missingRelations?

fanglong-alliance replied:
ah, nvm - looks like wont happen w/ usage below

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts:
```diff
@@ -85,45 +309,322 @@ export type ViewModelInterface<
+        }
+
+        for (const fieldName of nonRelatedFieldNames) {
+            // TODO: Unclear to me if this needs to call a method on the Field on not. Revisit this.
```
just checkimng that this todo would stay as todo for now?

davecoates replied:
i'll remove it - it's been there since original version

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts:
```diff
@@ -85,45 +309,322 @@ export type ViewModelInterface<
+                    if (field.many && !assignedData[key]) {
+                        assignedData[key] = [];
+                    }
+                    // TODO: Make doing this part of interface rather than special case?
```
(todo)

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts:
```diff
@@ -85,45 +309,322 @@ export type ViewModelInterface<
+                        const pk = Array.isArray(pkOrPks) ? pkOrPks.join(', ') : pkOrPks;
+                        console.warn(
+                            `Related field ${name} was created from nested object that had a different id to the source field name ${sourceFieldName}: ${data[sourceFieldName]} !== ${pk}. ${pk} has been used for both.`
+                        );
```
can we add the model.name to this warning? to make it a bit clearer

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts:
```diff
@@ -132,33 +633,90 @@ export type ViewModelInterface<
+        const { pkFieldName } = this._model;
+        if (Array.isArray(pkFieldName)) {
+            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+            // @ts-ignore
```
can use some text here explaining why (thought we're doing this everywhere else?)

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/fieldUtils.ts:
```diff
@@ -0,0 +1,244 @@
+ * For a model expand the fields according to the rules for the '*' specification.
+ *
+ * This includes all field names apart from relation fields (but it does include the
+ * relation `sourceFieldName`).
```
^but

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/fieldUtils.ts:
```diff
@@ -0,0 +1,244 @@
+                    const field = modelClass.getField(relatedFieldName);
+                    if (!(field instanceof BaseRelatedViewModelField)) {
+                        throw new InvalidFieldError(
+                            `Field '${relatedFieldName}' does not extend BaseRelatedViewModelField.`
```
also print the model name here?

fanglong-alliance commented on js-packages/@prestojs/viewmodel/src/fieldUtils.ts:
```diff
@@ -0,0 +1,244 @@
+        f.sort();
+        // See if it already exists
+        for (const viewModelFieldPaths of this.cache.values()) {
+            if (viewModelFieldPaths.key === f.join(CACHE_KEY_FIELD_SEPARATOR)) {
```
consider to do a join outside? just to make it a tiny bit faster

davecoates committed:
fix: minor changes from review

davecoates committed:
fix: fix matching on subset cache fields

davecoates committed:
fix: add getter on old cache attribute, changelog

davecoates committed:
fix: getList overload

Add overload to getList such that this works:

```
getList: <T extends SomeModel>(
    cache: ViewModelCache<typeof SomeModel>,
    records: T[] | null,
): T[] => {
    if (!records) {
        return [];
    }
    return cache.getList(result, true);
}
```

Without this overload the return type `T[]` won't match getList
(it will be PartialViewModel).

davecoates committed:
doc: changelog
